// ts-mls/dist/src/message.js
import { flatMapDecoder, mapDecoder, mapDecoders } from "./codec/tlsDecoder.js";
import { contramapBufferEncoders } from "./codec/tlsEncoder.js";
import { groupInfoDecoder, groupInfoEncoder } from "./groupInfo.js";
import { keyPackageDecoder, keyPackageEncoder } from "./keyPackage.js";
import { privateMessageDecoder, privateMessageEncoder } from "./privateMessage.js";
import { protocolVersionDecoder, protocolVersionEncoder } from "./protocolVersion.js";
import { publicMessageDecoder, publicMessageEncoder } from "./publicMessage.js";
import { welcomeDecoder, welcomeEncoder } from "./welcome.js";
import { wireformatDecoder, wireformatEncoder, wireformats } from "./wireformat.js";
export const mlsPublicMessageEncoder = contramapBufferEncoders([wireformatEncoder, publicMessageEncoder], (msg) => [msg.wireformat, msg.publicMessage]);
export const mlsWelcomeEncoder = contramapBufferEncoders([wireformatEncoder, welcomeEncoder], (wm) => [wm.wireformat, wm.welcome]);
export const mlsPrivateMessageEncoder = contramapBufferEncoders([wireformatEncoder, privateMessageEncoder], (pm) => [pm.wireformat, pm.privateMessage]);
export const mlsGroupInfoEncoder = contramapBufferEncoders([wireformatEncoder, groupInfoEncoder], (gi) => [gi.wireformat, gi.groupInfo]);
export const mlsKeyPackageEncoder = contramapBufferEncoders([wireformatEncoder, keyPackageEncoder], (kp) => [kp.wireformat, kp.keyPackage]);
export const mlsMessageContentEncoder = (mc) => {
    switch (mc.wireformat) {
        case wireformats.mls_public_message:
            return mlsPublicMessageEncoder(mc);
        case wireformats.mls_welcome:
            return mlsWelcomeEncoder(mc);
        case wireformats.mls_private_message:
            return mlsPrivateMessageEncoder(mc);
        case wireformats.mls_group_info:
            return mlsGroupInfoEncoder(mc);
        case wireformats.mls_key_package:
            return mlsKeyPackageEncoder(mc);
    }
};
export const mlsMessageContentDecoder = flatMapDecoder(wireformatDecoder, (wireformat) => {
    switch (wireformat) {
        case wireformats.mls_public_message:
            return mapDecoder(publicMessageDecoder, (publicMessage) => ({ wireformat, publicMessage }));
        case wireformats.mls_welcome:
            return mapDecoder(welcomeDecoder, (welcome) => ({ wireformat, welcome }));
        case wireformats.mls_private_message:
            return mapDecoder(privateMessageDecoder, (privateMessage) => ({ wireformat, privateMessage }));
        case wireformats.mls_group_info:
            return mapDecoder(groupInfoDecoder, (groupInfo) => ({ wireformat, groupInfo }));
        case wireformats.mls_key_package:
            return mapDecoder(keyPackageDecoder, (keyPackage) => ({ wireformat, keyPackage }));
    }
});
/** @public */
export const mlsMessageEncoder = contramapBufferEncoders([protocolVersionEncoder, mlsMessageContentEncoder], (w) => [w.version, w]);
/** @public */
export const mlsMessageDecoder = mapDecoders([protocolVersionDecoder, mlsMessageContentDecoder], (version, mc) => ({ ...mc, version }));
